### ISSUE TITLE
Animation export logic cause a huge impact on performance when sampling is enabled

### ISSUE THREAD

**Issue by HungryProton** [KhronosGroup/glTF-Blender-IO#647]
When **Always sample animations** is active, the exporter seems to sample the animations way too many times than what's actually needed.

_I'm not familiar with the code and my experience with Blender addons is rather low but I'll try my best to explain what I found._

Breaking down the export code, we have the following situation :

+ Starting from `gltf2_blender_gather_animations.py`, the exporter loop through every active animations
+ For each animation, we create a `gltf2_io animation object` on `line 82`
+ To create this object, we need all the channels so we move to `gltf2_blender_gather_animation_channels.py` **where I believe the issue begins**
+ On `line 53` for each bone, we call `__gather_animation_channel` **three times**
+ In order to get the `AnimationChannels on line 109`, we need a sampler so we move to `gltf2_blender_gather_animation_samplers.py`
+ We call `__gather_input on line 61` and `__gather_output on line 64`, both of them call the **heaviest function** of all : `gather_keyframes` defined in `gltf2_blender_gather_animation_sampler_keyframes.py on line 116`
+ I'm not 100% sure of what this function does exactly, but since all the constraints for the given bone have to be solved before adding it to the keyframe, I assume the entire pose tree is rebuilt every time.

Let's take a concrete example. I have an armature composed of 60 bones and a single 24 frames long animation. When I export the object this happens : 

``` 
For each bones in the 60 bones armature
  For each ["location", "rotation_quaternion", "scale"]
     call __gather_input // which calls gather_keyframes
     call __gather_output // which also calls gather_keyframes
```

On my system (Ryzen 5 1600), `gather_keyframes` takes about **90ms** to complete.
`90ms * 60 bones * 3 parameters * 2 calls =  32 seconds` for a single 24 frames animation.
Compared to the Collada exporter for example, exporting 32 animations with the same armature takes 4 seconds.

Again I'm no expert but I'm sure we can optimize this. We could already cut that time by half by reusing that `gather_keyframes` call in both `__gather_input` and `__gather_output`.

Ideally, for a 24 frames animation, the sampling should only happen 24 times, not 360 so it should probably move higher in the chain but it would require to refactor a significant amount of code.

**Comment by pgruenbacher** [comment]
yea I see the same issue.

**Comment by julienduroure** [comment]
Hello,

gather_keyframes is cached, so data are calculated only once.

But you're right, this is a know issue that frames are evaluated for each bone. This is already on my TODO list to cache info for each bone, and go frame by frame only once.

**Comment by Master-J** [comment]
I noticed similar export times. The issue is even more noticeable when exporting models with multiple animations, the export times can quickly take up to mutliple minutes.

Some measurements i did (45 bones armature)
- one animation (160 frames) : 38 seconds
- 16 animations (varrying length) : more than 40 minutes (thought blender was crashing so i killed it before it finishes the export)

**Comment by HungryProton** [comment]
@julienduroure Thank you for your quick answer! :smiley: 
If there is a cache then I'm not sure it's working as intended.
I added a few print statements :
![image](https://user-images.githubusercontent.com/52043844/64080360-9165d500-ccf3-11e9-843c-c2f2268006ef.png)
The function is still called a huge amount of times. I tried to uncomment the @cached property on line 110 and the total export time went from 32s to 18s. I'll keep searching.

**Comment by pgruenbacher** [comment]
I  made a PR that helps at least with not export all the bones for every animation, which is a huge issue with small animations on armature with many bones. 
https://github.com/KhronosGroup/glTF-Blender-IO/pull/648

**Comment by julienduroure** [comment]
Thanks for the patch. Unfortunately, this PR can't be accepted. 
Export only bones related will break most of animation export, when deformations bones are not keyed. (And not sure joint structure is handle correctly in your patch (not verified))

The solution is to export only deformation bones, the patch is WIP on my side

**Comment by julienduroure** [comment]
@HangryProton 

Export only Def bones is planned, and listed here #260

Function are called for every bone. That what will be change baking frame by frame only once. Current caching avoid calling these function multiple times for every bone (that is needed to get input / output for example)

**Comment by julienduroure** [comment]
A first step will be to cache bone matrix for all bones, to avoid to set frame for each bone. Will be manage in #654

**Comment by julienduroure** [comment]
This will be solved by #664 after merging it.
If you want me to have a look on your file to test perf improvement, please upload it

**Comment by julienduroure** [comment]
Fixed by merging #664

**Comment by Master-J** [comment]
The problem seems fixed, tested the fix with the same model as before (45 bones armature, 16 animations). The export time went from over an hour to a minute long.

Thanks for the fix.
![Capture](https://user-images.githubusercontent.com/54776415/64761381-3d04f580-d53c-11e9-879e-7ef471d689c3.PNG)
